cpp/include/cudf/utilities/error.hpp

```diff
@@ -104,9 +104,11 @@ inline void throw_cuda_error(cudaError_t error, const char* file,
                   cudaGetErrorName(error) + " " + cudaGetErrorString(error)});
 }
 
-inline void check_stream(cudaStream_t stream, const char* file,
-                         unsigned int line) {
+template <bool sync_on_stream>
+inline std::enable_if_t<sync_on_stream, void>
+check_cuda_error(char const* file, unsigned int line, cudaStream_t stream=0) {
```

Comment:
This is more complex than what I had intended. I meant something like this:
```
template <bool sync_on_stream>
inline void check_cuda_error(char const* file, unsigned int line, cudaStream_t stream=0) {

   if(sync_on_stream)
      ...
   else
      ...
```

Since `sync_on_stream` is known at compile time, the compiler will optimize away one of the two code paths. Before when `sync_on_stream` was just a runtime parameter, I wasn't as confident that this will happen without it being a `constexpr` function.